application/common/components/MfaBackendWebAuthn.php

```diff
@@ -125,29 +135,65 @@ public function verify(int $mfaId, $value, string $rpOrigin = '')
             $mfa->external_uuid
         );
 
+
         if (!is_array($value)) {
             $value = Json::decode($value);
             if ($value == null) {
                 throw new ServerErrorHttpException("Missing data or unable to decode as JSON", 1638447364);
             }
         }
 
-        if ($mfa->verified === 1) {
-            return $this->client->webauthnValidateAuthentication($headers, $value);
-        } else {
-            $results = $this->client->webauthnValidateRegistration($headers, $value);
-            if (isset($results['key_handle_hash'])) {
-                $mfa->verified = 1;
-                $mfa->key_handle_hash = $results['key_handle_hash'];
-                if (! $mfa->save()) {
-                    throw new ServerErrorHttpException(
-                        "Unable to save WebAuthn record after verification. Error: " . print_r($mfa->getFirstErrors(), true)
-                    );
-                }
-                return true;
+        // If the verifyType is not registration, finish the login process.
+        if ($verifyType != Mfa::VERIFY_REGISTRATION) {
+            // Ensure there is at least one verified MfaWebauthn for that Mfa
+            $webauthnCount = MfaWebauthn::find()->where(['mfa_id' => $mfa->id,'verified' => 1])->count();
+            if ($webauthnCount < 1) {
+                throw new NotFoundHttpException("No verified MFA Webauthn record found for MFA ID: " . $mfa->id, 1659637860);
             }
+            return $this->client->webauthnValidateAuthentication($headers, $value);
+        }
+
+        // Assume a new webauthn was requested and finish its registration process
+        $results = $this->client->webauthnValidateRegistration($headers, $value);
+
+        if (! isset($results['key_handle_hash'])) {
             return false;
         }
+
+        $label = $mfa->getReadableType();
+        self::createWebauthn($mfa, $results['key_handle_hash']);
+
+        return true;
+    }
+
+
+
+    /**
+     * @param string $label
+     * @param string $keyHandleHash
+     * @return void
+     * @throws ServerErrorHttpException
+     */
+    public static function createWebauthn(Mfa $mfa, string $keyHandleHash): void {
```

Comment:
Sorry, I was wrong about moving this `createWebauthn()` function down into this `MfaBackendWebAuthn` class. The creating of MfaWebauthn records really does belong in that MfaWebauthn class. This "`MfaBackendWebAuthn` is really for handling interactions with the MFA API that are related to webauthn.